### PR TITLE
Add customizable expungeStrategy

### PR DESCRIPTION
**Pull request by ANeumann82**
This would allow us to use an expungeStrategy in Metronome that just expunges finished/failed instances regardless of the current goal

JIRA-Issues: DCOS-50471

### CODE REVIEW

**Comment by jeschkies** [comment]
Hm, this code is less disruptive than I thought. However, I like the concept of a life cycle policy much better as it opens the door for number of retries etc. Instead of expunging we would simply change the goal to decommissioned. What do you think?

**Comment by ANeumann82** [comment]
I'm not sure if that would work correctly.

This code is in the mesosUpdate, i.e. it's executed when we receive an update from mesos.
If we don't issue the Expunge here, but only change the goal, we need to trigger the Expunge at some other point, as we're probably not going to receive another update from mesos regarding this instance.

Having a rescheduleStrategy on the runSpec instead of the ExpungeStrategy here might be a better way, but would involve quite a bit more work, including a migration for RunSpecs. I'd rather not do that in this PR, but in a new one that has fleshed out feature requirements

File: src/main/scala/mesosphere/marathon/core/instance/update/InstanceUpdater.scala [hunk 1]
```diff
@@ -10,10 +10,37 @@ import mesosphere.marathon.core.task.update.TaskUpdateEffect
 import mesosphere.marathon.state.{Timestamp, UnreachableEnabled}
 import org.apache.mesos.{Protos => MesosProtos}
 
+trait InstanceExpungeStrategy {
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/core/instance/update/InstanceUpdater.scala]: I guess this is an implementation detail then. However, I'd prefer a rescheduling strategy and have the definition for `schouledBeExpunged` be inferred from that. There is no need to change the goal. As long as we agree not to attach an expunge strategy to the run spec I'm fine with this change.

Comment (ANeumann82) [reply src/main/scala/mesosphere/marathon/core/instance/update/InstanceUpdater.scala]: Yeah, we should not have the expungeStrategy on the runSpec, a rescheduling strategy makes a lot more sense there. 
And I think having the custom ExpungeStrategies here, implementing different ReschedulingStrategies per runSpec gets a lot easier.

**Pull request closed without merging**